### PR TITLE
Feature protocols/event schemas

### PR DESCRIPTION
**Pull request by SumanthPuram**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by SumanthPuram** [comment]
Test setup:
1. Create 2 sources
2. Send events to both the sources, also vary the payload. 

Test cases performed:
1. For the same event payload that is sent again to the same source, only the last_seen should be updated.
2. Same event sent to 2 different sources, should result in 2 different event_types
3. If the payload changes, a new schema version should be created
4. On restart of the server, application should keep the state.

File: protocols/protocols.go [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+func (manager *ProtocolManagerT) setupTables() {
+	createEventTypesSQL := fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s (
+		id SERIAL PRIMARY KEY,
+		uuid VARCHAR(36) NOT NULL,
+		write_key VARCHAR(32) NOT NULL,
+		ev_type TEXT NOT NULL,
+		event_identifier TEXT NOT NULL DEFAULT '',
+		created_at TIMESTAMP NOT NULL DEFAULT NOW()
+	)
+	`, EVENT_TYPES_TABLE)
```
Comment (psrikanth88) [inline protocols/protocols.go:452]: names are a bit confusing. there is a event type `track` and also an event type `Track + App Opened` etc. Consider renaming it?

File: protocols/protocols.go [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+func (manager *ProtocolManagerT) handleEvent(writeKey string, event EventT) {
+	evType := event["type"].(string)
+	eventIdentifier := ""
+	if evType == "track" {
+		eventIdentifier = event["event"].(string)
+	} else if evType == "page" {
+		eventIdentifier = event["name"].(string)
+	} else if evType == "screen" {
+		eventIdentifier = event["name"].(string)
+	}
```
[the remark this answers is not in the document]
Comment (SumanthPuram) [reply protocols/protocols.go:224]: Done

File: protocols/protocols.go [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+
+	//TODO: Create locks on every event_type to improve scaling this
+	manager.eventTypeLock.Lock()
+	manager.schemaVersionLock.Lock()
+	defer manager.eventTypeLock.Unlock()
+	defer manager.schemaVersionLock.Unlock()
+
+	eventType := manager.eventTypeMap[writeKey][evType][eventIdentifier]
```
[the remark this answers is not in the document]
Comment (SumanthPuram) [reply protocols/protocols.go]: Thanks for this catch. I missed it.

File: protocols/protocols.go [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+
+	eventMap := map[string]interface{}(event)
+
+	flattenedEvent, err := flatten.Flatten((eventMap), "", flatten.DotStyle)
+
+	if err != nil {
+		fmt.Println(err)
+		panic("Failed to flatten the event")
+	} else {
+		fmt.Println(flattenedEvent)
```
[the remark this answers is not in the document]
Comment (SumanthPuram) [reply protocols/protocols.go]: Done

File: protocols/protocols.go [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+		finalSchema := make(map[string]string)
+		keys := make([]string, 0, len(finalSchema))
```
[the remark this answers is not in the document]
Comment (SumanthPuram) [reply protocols/protocols.go]: Fixed

File: protocols/protocols.go (hunk 1 above)
Comment (SumanthPuram) [reply protocols/protocols.go]: Renamed event_types to event_models

File: gateway/gateway.go [hunk 6]
```diff
@@ -804,6 +808,10 @@ func (gateway *HandleT) StartWebHandler() {
 		srvMux.HandleFunc("/v1/webhook", gateway.stat(gateway.webhookHandler.RequestHandler))
 	}
 
+	// Protocols
+	srvMux.HandleFunc("/protocols/get-event-models", gateway.protocolManager.GetEventModels)
+	srvMux.HandleFunc("/protocols/get-event-versions", gateway.protocolManager.GetEventVersions)
```
Comment (psrikanth88) [inline gateway/gateway.go:813]: remove get in the URL's and also make it GET only in protocols_api.go ? In future, we might need to also support POST requests to register event models through API right?

Comment (SumanthPuram) [reply gateway/gateway.go]: Done

**Pull request merged**